modules/externalipam/ipam_api.rb

```diff
@@ -0,0 +1,389 @@
+require 'proxy/validations'
+require 'externalipam/ipam_helper'
+require 'externalipam/ipam_validator'
+require 'externalipam/dependency_injection'
+
+module Proxy::Ipam
+  # Generic API for External IPAM interactions
+  class Api < ::Sinatra::Base
+    extend Proxy::Ipam::DependencyInjection
+    include ::Proxy::Log
+    helpers ::Proxy::Helpers
+    include Proxy::Ipam::IpamHelper
+    include Proxy::Ipam::IpamValidator
+
+    inject_attr :externalipam_client, :client
+
+    # Gets the next available IP address based on a given External IPAM subnet
+    #
+    # Inputs:   1. address:          Network address of the subnet(e.g. 100.55.55.0)
+    #           2. prefix:           Network prefix(e.g. 24)
+    #           3. group(optional):  The External IPAM group
+    #
+    # Returns:
+    #   Response if success:
+    #   ======================
+    #     Http Code:     200
+    #     JSON Response:
+    #       "100.55.55.3"
+    #
+    #   Response if missing parameter(e.g. 'mac')
+    #   ======================
+    #     Http Code:     400
+    #     JSON Response:
+    #       {"error": ["A 'mac' address must be provided(e.g. 00:0a:95:9d:68:10)"]}
+    #
+    #   Response if no free ip's available
+    #   ======================
+    #     Http Code:     404
+    #     JSON Response:
+    #       {"error": "There are no free IP's in subnet 100.55.55.0/24"}
+    get '/subnet/:address/:prefix/next_ip' do
+      content_type :json
+
+      begin
+        validate_required_params!([:address, :prefix], params)
+
+        mac_param = params[:mac]
+        mac = validate_mac!(params[:mac]) unless mac_param.nil? || mac_param.empty?
+        cidr = validate_cidr!(params[:address], params[:prefix])
+        group_name = get_request_group(params)
+
+        next_ip = provider.get_next_ip(mac, cidr, group_name)
+        halt 404, { error: ERRORS[:no_free_ips] }.to_json if next_ip.nil?
+        next_ip.to_json
```

Comment:
One nasty thing I missed during my initial review - it is not a good practice to return any other values than hash in JSON. It is not extensible, if you want to add another element with this design (when it is actually a JSON-encoded string) you cannot do that easily. I would suggest something like this:

```suggestion
        {ip: next_ip}.to_json
```

A change in Foreman core will be required to fix this.

Reply:
Will be submitting a PR on the foreman side as well